File: c#/snail.cs
public class SnailSolution
{
    public static int[] Snail(int[][] array)
    {
        var traverser = new Traverser(array);
        int[] result = new int[array.Length * array[0].Length];

        int iteration = 0;
        while (!traverser.IsDone())
        {
            var position = traverser.Next();
            result[iteration] = array[position.y][position.x];
            iteration++;
        }
        return result;
    }
}

enum Direction
{
    RIGHT = 1,
    DOWN,
    LEFT,
    UP
}

class Traverser
{
    private readonly DirectionTracker directionTracker;
    private Direction direction;
    private Limit limit;
    private Position position;
    private int iteration;
    private readonly int maxIteration;

    public Traverser(int[][] array)
    {
        limit = new Limit(array.Length - 1, array[0].Length - 1);
        directionTracker = new DirectionTracker();
        maxIteration = array.Length * array[0].Length;
        iteration = 0;
    }

    public bool IsDone()
    {
        return maxIteration <= iteration;
    }

    public Position Next()
    {
        iteration++;
        switch (direction)
        {
            default:
                direction = directionTracker.Next();
                position = new Position { x = 0, y = 0 };
                break;
            case Direction.RIGHT:
                if (position.x < limit.Right)
                {
                    position.x++;
                }
                else
                {
                    limit.Top++;
                    position.y++;
                    direction = directionTracker.Next();
                }
                break;
            case Direction.DOWN:
                if (position.y < limit.Bottom)
                {
                    position.y++;
                }
                else
                {
                    limit.Right--;
                    position.x--;
                    direction = directionTracker.Next();
                }
                break;
            case Direction.LEFT:
                if (position.x > limit.Left)
                {
                    position.x--;
                }
                else
                {
                    limit.Bottom--;
                    position.y--;
                    direction = directionTracker.Next();
                }
                break;
            case Direction.UP:
                if (position.y > limit.Top)
                {
                    position.y--;
                }
                else
                {
                    limit.Left++;
                    position.x++;
                    direction = directionTracker.Next();
                }
                break;
        }
        return position;
    }
}

class DirectionTracker
{
    private Direction currentDirection;
    public Direction Next()
    {
        switch (currentDirection)
        {
            default:
            case Direction.UP:
                currentDirection = Direction.RIGHT;
                break;
            case Direction.RIGHT:
                currentDirection = Direction.DOWN;
                break;
            case Direction.DOWN:
                currentDirection = Direction.LEFT;
                break;
            case Direction.LEFT:
                currentDirection = Direction.UP;
                break;
        }

        return currentDirection;
    }
}

struct Position
{
    public int x;
    public int y;
}

struct Limit
{
    public Limit(int verticalLimit, int horizontalLimit)
    {
        Top = 0;
        Right = horizontalLimit;
        Left = 0;
        Bottom = verticalLimit;
    }

    public int Top;
    public int Right;
    public int Left;
    public int Bottom;
}

File: python/snail.py
def snail(array):
    myList = []
    n = len(array)
    i = 0
    #left, right, up, down
    bounds = [0,n-1, 0, n-1]
    #right, down, left, up
    direction = [1,0,0,0]
    inc = 0
    
    if n ==1:return array[0]
    
    while i != n*n:
        x = direction.index(1)
        if x == 0:
            myList.append(array[bounds[2]][inc])
            if inc == bounds[1]:
                direction[0] = 0
                direction[1] = 1
                bounds[2] +=1
                inc = bounds[2]
            else: inc+=1
        if x == 1:
            myList.append(array[inc][bounds[1]])
            if inc == bounds[3]:
                direction[1] = 0
                direction[2] = 1
                bounds[1] -= 1
                inc = bounds[1]
            else: inc += 1
        if x == 2:
            myList.append(array[bounds[3]][inc])
            if inc == bounds[0]:
                direction[2] = 0
                direction[3] = 1
                bounds[3] -= 1
                inc = bounds[3]
            else: inc -= 1
        if x == 3:
            myList.append(array[inc][bounds[0]])
            if inc == bounds[2]:
                direction[3] = 0
                direction[0] = 1
                bounds[0] += 1
                inc = bounds[0]
            else: inc -= 1
        i+=1
    return(myList)